ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,551 @@
+  -- still in the race to serve that candidate.
+  WithDeclined
+    peerInfo
+    (Maybe (ChainSuffix header, [(ChainSuffix header, peerInfo)]))
+selectTheCandidate
+  FetchDecisionPolicy {plausibleCandidateChain}
+  currentChain =
+    separateDeclinedAndStillInRace
+      -- Select the suffix up to the intersection with the current chain. This can
+      -- eliminate candidates that fork too deep.
+      . selectForkSuffixes currentChain
+      -- Filter to keep chains the consensus layer tells us are plausible.
+      . filterPlausibleCandidates plausibleCandidateChain currentChain
+    where
+      -- Write all of the declined peers, and find the longest candidate
+      -- fragment if there is any.
+      separateDeclinedAndStillInRace ::
+        [(FetchDecision (ChainSuffix header), peerInfo)] ->
+        WithDeclined peerInfo (Maybe (ChainSuffix header, [(ChainSuffix header, peerInfo)]))
+      separateDeclinedAndStillInRace decisions = do
+        let (declined, inRace) = partitionEithers
+              [ bimap ((,p)) ((,p)) d | (d, p) <- decisions ]
+        tell (List declined)
+        case inRace of
+          [] -> pure Nothing
+          _ : _ -> do
+            let chainSfx = fst $
+                  List.maximumBy (compare `on` (headBlockNo . getChainSuffix . fst))  inRace
+            pure $ Just (chainSfx, inRace)
+
+-- | Given _the_ candidate fragment to sync from, and a list of peers (with
+-- their corresponding candidate fragments), choose which peer to sync _the_
+-- candidate fragment from.
+--
+-- We first filter out all the peers that cannot even serve a reasonable batch
+-- of _the_ candidate fragment, and then we choose the first one according to
+-- the ordering passed as argument.
+--
+-- PRECONDITION: The set of peers must be included in the peer order queue.
+-- PRECONDITION: The given candidate fragments must not be empty.
+selectThePeer ::
+  forall header peer extra.
+  HasHeader header =>
+  -- | The candidate fragment that we have selected to sync from, as suffix of
+  -- the immutable tip.
+  CandidateFragments header ->
+  -- | Association list of candidate fragments (as suffixes of the immutable
+  -- tip) and their associated peers.
+  [(ChainSuffix header, PeerInfo header peer extra)] ->
+  WithDeclined
+    (PeerInfo header peer extra)
+    (Maybe (ChainSuffix header, PeerInfo header peer extra))
+selectThePeer
+  theFragments
+  candidates = do
+    -- Create a fetch request for the blocks in question. The request has exactly
+    -- 1 block. It will only be used to choose the peer to fetch from, but we will
+    -- later craft a more refined request for that peer. See [About the gross
+    -- request] in the module documentation. Because @theFragments@ is not
+    -- empty, @grossRequest@ will not be empty.
+    let firstBlock = FetchRequest . map (AF.takeOldest 1) . take 1 . filter (not . AF.null)
+        (grossRequest :: FetchRequest header) = firstBlock $ snd theFragments
```

Comment:
> Because `theFragments` is not empty, `grossRequest` will not be empty.

What if all fragments are filtered out by `filter (not . AF.null)`? Should we abort early and return `Nothing`? Or is this not possible?

Reply:
The comment two lines above claims `theFragments` to be non-empty. It is produced by `dropAlreadyFetchedBlocks` which is intended to yield a non-empty fragment.

Reply:
Ok, so:

 - `snd theFragments` is not empty because it is computed via `dropAlreadyFetchedBlocks` which calls `dropAlreadyFetched` which declines when the list of fragments is empty.
   https://github.com/IntersectMBO/ouroboros-network/blob/5e4515e8bb5dbac44b779fd5d63b2e6eb06f3fc1/ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/Deadline.hs#L555-L556
 - Moreover, every element of `snd theFragments` is a non-empty fragment because `dropAlreadyFetched` calls `filterWithMaxSlotNo` which calls [`filterWithStop`](https://ouroboros-network.cardano.intersectmbo.org/ouroboros-network-api/Ouroboros-Network-AnchoredFragment.html#v:filterWithStop) which returns a list of non-empty fragments.

Therefore, I think the `filter (not . AF.null)` will never filter anything out, which confirms that `grossRequest` is always non-empty :+1:. We could remove `filter (not . AF.null)` therefore :thinking:

Reply:
It would be fine to drop the filter and put an assertion, I guess.